presto-base-jdbc/src/main/java/io/prestosql/plugin/jdbc/CachingJdbcClient.java

```diff
@@ -346,39 +357,77 @@ public String quoted(RemoteTableName remoteTableName)
         return delegate.getTableScanRedirection(session, tableHandle);
     }
 
+    protected ColumnsCacheKey buildColumnsCacheKey(ConnectorSession session, SchemaTableName tableName)
+    {
+        return new ColumnsCacheKey(JdbcIdentity.from(session), getSessionProperties(session), tableName);
+    }
+
+    private TableHandleCacheKey buildTableHandleCacheKey(ConnectorSession session, SchemaTableName schemaTableName)
+    {
+        return new TableHandleCacheKey(JdbcIdentity.from(session), getSessionProperties(session), schemaTableName);
+    }
+
+    private TableNamesCacheKey buildTableNamesCacheKey(ConnectorSession session, Optional<String> schema)
+    {
+        return new TableNamesCacheKey(JdbcIdentity.from(session), getSessionProperties(session), schema);
+    }
+
     private void invalidateSchemasCache()
     {
         schemaNamesCache.invalidateAll();
     }
 
-    private void invalidateTablesCaches()
+    private void invalidateTablesCaches(ConnectorSession session, SchemaTableName table)
     {
-        columnsCache.invalidateAll();
-        tableHandleCache.invalidateAll();
-        tableNamesCache.invalidateAll();
+        invalidateColumnsCache(session, table);
+        invalidateTableCache(session, table);
+    }
+
+    private void invalidateTableCache(ConnectorSession session, SchemaTableName table)
+    {
+        invalidateCache(tableNamesCache, cacheKey -> cacheKey.schemaName.equals(Optional.of(table.getSchemaName())));
+        invalidateCache(tableHandleCache, cacheKey -> cacheKey.tableName.equals(table));
     }
 
     private void invalidateColumnsCache(ConnectorSession session, SchemaTableName table)
     {
-        columnsCache.invalidate(new ColumnsCacheKey(JdbcIdentity.from(session), table));
+        invalidateCache(columnsCache, cacheKey -> cacheKey.table.equals(table));
+    }
+
+    private Map<String, Object> getSessionProperties(ConnectorSession session)
+    {
+        return sessionPropertiesProviders.stream()
+                .map(property -> Map.entry(property.getName(), session.getProperty(property.getName(), property.getJavaType())))
+                .collect(toImmutableMap(Map.Entry::getKey, Map.Entry::getValue));
     }
 
-    private static final class ColumnsCacheKey
+    private static <T, V> void invalidateCache(Cache<T, V> cache, Predicate<T> filterFunction)
+    {
+        Set<T> cacheKeys = ImmutableMap.copyOf(cache.asMap()).keySet().stream()
```

Comment:
do you really need `copyOf`?

Reply:
Nope. Just being defensive here.

Reply:
Could you please remove it? You don't expect 'changes from outside' and everything is under your control.